tips/tip-0008/tip-0008.md

```diff
@@ -51,9 +51,26 @@ The state of the VM is fully determined by a bunch of things, divisible into the
 
 The VM is initialized with the program hash on the bottom of its opstack. By making these five elements **read-only**, we get program integrity from opstack integrity and collision-resistance.
 
-The only way in which the program hash could be overwritten is through `swap` instructions when the opstack size is 21 or less. 
+The only way in which the program hash could be overwritten is through `swap`, `pick`, and `place` instructions when the opstack size is 21 or less. 
 
-**Constraints.** Todo.
+**Constraints.** To prevent the program hash from being overwritten by stack manipulation instructions when the opstack size is 21 or less, we need to ensure that these operations cannot target the program hash elements (st11 through st15) when the stack is shallow enough to expose them.
+
+Let $n$ be the current opstack size and let $i$ be the target index. The constraint is:
+
+$$\text{stack\_manip\_indicator} \cdot \text{indicator}_{n \leq 21} \cdot \text{indicator}_{i \geq 11} = 0$$
+
+where:
+- $\text{stack\_manip\_indicator}$ is 1 when executing a `swap`, `pick`, or `place` instruction, 0 otherwise
+- $\text{indicator}_{n \leq 21}$ is 1 when $n \leq 21$ (stack is shallow enough to expose program hash), 0 otherwise
+- $\text{indicator}_{i \geq 11}$ is 1 when $i \geq 11$ (targeting program hash elements), 0 otherwise
+
+This constraint ensures that when the stack size is 21 or less ($n \leq 21$), any stack manipulation operation targeting program hash elements ($i \geq 11$) will be invalidated by the constraint evaluation.
+
+The indicator function $\text{indicator}_{n \leq 21}$ can be implemented as an arithmetic circuit of low degree using the fact that $n$ is bounded by the maximum stack size. Specifically, we can use the constraint:
+
+$$\text{indicator}_{n \leq 21} = \prod_{k=22}^{n_{max}} (n - k)$$
```

Comment:
I think it is correct now, but you've introduced another drawback: the maximum stack size is actually $P-1 = 2^{64} - 2^{32}$ which is *huge* when it is supposed to be a bound on the degree of an AIR constraint.